### DISCUSSION TITLE
尝试用 Rust 实现了一个 Waline 的迷你版本，过程中有几个吐槽的地方 || I tried to implement a mini version of Waline using Rust. There were several complaints in the process.

### DISCUSSION THREAD

**Discussion by JQiue** [walinejs/waline#2868] [General]
- `Get /api/comment`是最坑爹的一个设计，一个接口要靠字段来区分普通用户和管理员的处理逻辑
- `2fa`数据库这个字段也是设计不好的地方，因为有一些orm框架会从数据库直接生成某个结构的成员，结果就是并不支持数字开头需要手动处理
- api文档并没有写清楚穿参是qs还是param或者body，这必须通过浏览器发请求确认
- 响应体中的code错误以为只有1000，就把所有错误全部搞成1000了，结果通过api接口工具测试时偶然发现居然还有1001，api文档也没有写清楚有哪些错误码，所以照着api文档实现不如从浏览器逆向以及看源码

---
- `Get /api/comment` is the most deceptive design. An interface relies on fields to distinguish the processing logic of ordinary users and administrators.
- The `2fa` database field is also poorly designed, because some ORM frameworks will directly generate members of a certain structure from the database. As a result, the beginning of a number is not supported and needs to be processed manually.
- The api document does not clearly state whether the parameter is qs, param or body. This must be confirmed by sending a request through the browser.
- The code error in the response body was thought to be only 1000, so I changed all the errors to 1000. When testing with the api interface tool, I accidentally discovered that there was 1001. The api document did not clearly state what error codes there were, so I followed the api. Document implementation is not as good as reverse engineering from the browser and looking at the source code

**Comment by lizheming** [comment]
1. 是说 `type=list` 么，这个是比较早期的实现了，当时主要是使用场景不一样，数据字段，分页以及父子评论结构等逻辑都不一样，就先简单来了。现在回过头看，前台也增加了管理员模式，其实两个是可以合并成一个，用更细粒度的功能参数区分。你可以说下你觉得比较好的方式是什么，可以一并参考下。
2. `2fa` 这个确实。当时定名的时候确实也是有纠结一下，主要是想选个比较简短的名称，最后定了它。后面可以改下，使用的用户应该也不多。
3. 因为 `1000` 和 `1001` 错误码都是底层框架 ThinkJS 返回的，所以没有特别的说明。实际上你不说我都不知道还有 1001 的返回。前端一般就是识别非0了，没有特别关注过具体的错误好。另外刚才搜索了下目前还有 `401` 和 `403` 相关的用户登录权限相关的错误。

---
1. Do you mean `type=list`? This was implemented relatively early. At that time, the main reason was that the usage scenarios were different, and the data fields, paging, parent-child comment structure and other logic were different, so I will simplify it first. Looking back now, the administrator mode has also been added to the front desk. In fact, the two can be merged into one and distinguished by more fine-grained functional parameters. You can tell me what you think is a better way, and you can refer to it together.
2. `2fa` This is true. I was really conflicted when I was deciding on the name, mainly because I wanted to choose a shorter name, and finally settled on it. It can be changed later, and there shouldn't be many users using it.
3. Because the `1000` and `1001` error codes are returned by the underlying framework ThinkJS, there is no special explanation. In fact, I wouldn’t have known there was a 1001 return unless you told me. The front-end generally only recognizes non-zero values, and does not pay special attention to specific errors. In addition, I just searched and there are currently `401` and `403` related errors related to user login permissions.

**Reply by JQiue** [reply]
`type=list`是管理员才能用的字段，和普通用户字段放到一起处理字段是否存在挺麻的，更建议开一个新的 api 专门用于管理员后台获取评论列表，同时保留原有的 api 不动，逐渐替换 client 请求保证迁移没问题，2fa 也是，数据库加一个 two_factor_auth 字段，保留原有的 2fa 字段，反正每次打开管理员后台时会发一个请求验证 2fa，在后端处理这个请求的同时把 2fa 的数据写到 two_factor_auth 里面，以后看逐渐废掉 2fa 就行了

---
`type=list` is a field that can only be used by administrators. It is awkward to put it together with ordinary user fields to handle whether the fields exist. It is more recommended to open a new API specifically for the administrator to obtain the comment list in the background, while retaining the original The api does not move, and client requests are gradually replaced to ensure that the migration is no problem. The same goes for 2fa. The database adds a two_factor_auth field and retains the original 2fa field. Anyway, every time the administrator background is opened, a request will be sent to verify 2fa, and this request will be processed on the backend. at the same time The data of 2fa is written into two_factor_auth. In the future, 2fa will be gradually abolished.

**Comment by lizheming** [comment]
另外，如果项目有开源的话可以发一下地址，我可以帮你在用户群里也推广下~

---
In addition, if the project is open source, you can send the address and I can help you promote it among the user group~

**Reply by JQiue** [reply]
是开源的，https://github.com/JQiue/waline-mini

---
It is open source, https://github.com/JQiue/waline-mini

**Comment by JQiue** [comment]
意外发现 leancloud 支持传可执行文件部署，试了一下可以直接上传可执行文件和 SQLite 文件，只要不重新部署就不会丢失数据（重新部署先导出数据之后在导入数据就行了），这意味着存储额度要比 leancloud 要高出不少

---
I accidentally discovered that leancloud supports executable file deployment. After trying it, I can directly upload executable files and SQLite files. As long as I don’t redeploy, I won’t lose the data (re-deployment, just export the data first and then import the data), which means that storage The limit is much higher than leancloud

**Reply by lizheming** [reply]
是想说tidb cloud?

---
Do you want to say tidb cloud?

**Reply by JQiue** [reply]
不是tidb，就是leancloud可以直接拉仓库部署，另外还有个shuttle也可以部署，官方提供域名且国内可访问

---
Either tidb or leancloud can directly pull the warehouse for deployment. There is also a shuttle that can also be deployed. The official domain name is provided and accessible in China

**Reply by lizheming** [reply]
不明觉厉，没看特别明白，哈哈哈哈~ 等你分享吧~

---
I don’t know how serious it is, I didn’t understand it clearly after watching it, hahahaha~ I’ll wait for you to share it~

**Reply by JQiue** [reply]
> 不明觉厉，没看特别明白，哈哈哈哈~ 等你分享吧~
> 
> I don’t know how serious it is, I didn’t understand it clearly after watching it, hahahaha~ I’ll wait for you to share it~

还有一点小小的问题，我在做数据迁移的时候发现，`GET /api/db`导出的数据，每条记录的时间居然是格式化后的本地时间字符串，而不是 UTC 时间，我看 waline `POST /api/db`导入数据又把这个字符串又转换成 UTC，为什么不统一吐 UTC 时间呢？

**Reply by lizheming** [reply]
1. 没有吐出 UTC 的时间，是因为 `GET /api/db` 这块没有做数据处理，直接用了从数据库 SELECT 出来的数据导出了。而默认 LeanCloud 返回数据的时候返回的就是格式化后的时间，所以其它数据库也沿用了。这也是为什么 `POST /api/db` 又重新格式化字符串的原因。
2. 这块其实是个弊端，比较容易出现前后端时区不一致的问题。所以后来 `GET /api/comment` 接口增加了一个 `time` 字段，专门用来返回时间戳的，`insertAt`, `createAt`, `updateAt` 对于前端算是废弃字段了。不过因为 `/api/db` 这个接口用的不多，所以这块的逻辑并没有同步过来，应该是要改一下的。

---
1. The UTC time is not spit out because there is no data processing in `GET /api/db`, and the data from SELECT from the database is directly exported. By default, when LeanCloud returns data, it returns the formatted time, so other databases also use it. This is why `POST /api/db` reformats the string again.
2. This is actually a drawback. It is easier to have inconsistent time zones between the front and back ends. So later, the `GET /api/comment` interface added a `time` field, which is specifically used to return timestamps. `insertAt`, `createAt`, and `updateAt` are considered obsolete fields for the front end. However, because the `/api/db` interface is not used much, the logic of this piece has not been synchronized and should be changed.
